src/storage/compression/compression.cpp

```diff
@@ -455,48 +456,80 @@ void IntegerBitpacking<T>::setValuesFromUncompressed(const uint8_t* srcBuffer, o
     // with a size of 32 * bitWidth bits,
     // or bitWidth 32-bit values (we cast the buffer to a uint32_t* later).
     auto chunkStart = (uint8_t*)getChunkStart(dstBuffer, posInDst, header.bitWidth);
-    auto posInChunk = posInDst % CHUNK_SIZE;
-    U chunk[CHUNK_SIZE];
-    // TODO(bmwinger): Now I'm a bit confused here. Why do we always need to unpack the chunk first?
-    //      Can't we just unpack the chunk if we need to modify only partial of it?
-    fastunpack(chunkStart, chunk, header.bitWidth);
-    for (offset_t i = 0; i < numValues; i++) {
-        auto value = ((T*)srcBuffer)[posInSrc + i];
-        // Null values will usually be 0, which will not be able to be stored if there is a non-zero
-        // offset However we don't care about the value stored for null values Currently they will
-        // be mangled by storage+recovery (underflow in the subtraction below)
-        KU_ASSERT((nullMask && nullMask->isNull(posInSrc + i)) ||
-                  canUpdateInPlace(std::span(&value, 1), metadata));
-        chunk[posInChunk] = (U)(value - (T)header.offset);
-        if (posInChunk + 1 >= CHUNK_SIZE && i + 1 < numValues) {
-            fastpack(chunk, chunkStart, header.bitWidth);
-            chunkStart = (uint8_t*)getChunkStart(dstBuffer, posInDst + i + 1, header.bitWidth);
-            fastunpack(chunkStart, chunk, header.bitWidth);
-            posInChunk = 0;
-        } else {
-            posInChunk++;
-        }
-    }
-    fastpack(chunk, chunkStart, header.bitWidth);
+    const size_t lastAlignedChunkStartIdx =
+        std::max(posInDst, (posInDst + numValues) - (posInDst + numValues) % CHUNK_SIZE);
+    if (lastAlignedChunkStartIdx > posInDst) {
+        auto posInChunk = posInDst % CHUNK_SIZE;
+        U chunk[CHUNK_SIZE];
+        // TODO(bmwinger): Now I'm a bit confused here. Why do we always need to unpack the chunk
+        // first?
+        //      Can't we just unpack the chunk if we need to modify only partial of it?
+        fastunpack(chunkStart, chunk, header.bitWidth);
```

Comment:
This change should be able to address the comment here. You can do an unaligned pack first (like the one at the end) and then this loop can be modified to remove the unpack and just pack the 32-value chunks (which should be able to be simplified; the looping is a little complicated because it starts and ends at arbitrary points in the chunk, but if its only being used for full chunks the values can just be copied from the srcBuffer into the temporary chunk minus the offset, and then packed, before moving to the next 32-value chunk).

Reply:
updated